File: FUNCIONAL(Haskell)/ReviewManagement.hs
module ReviewManagement where

import System.IO (appendFile, readFile)
import Data.List (isInfixOf, words)
import Text.Read (readMaybe) -- Para parsing seguro

-- Função para deixar uma avaliação
leaveReview :: IO ()
leaveReview = do
    putStrLn "Avalie a infraestrutura da sala (nota de 0 a 10):"
    infraRating <- getLine
    putStrLn "Avalie a comida (nota de 0 a 10):"
    foodRating <- getLine

    let reviewEntry = "Infraestrutura: " ++ infraRating ++ " | Comida: " ++ foodRating ++ "\n"
    appendFile "reviews.txt" reviewEntry
    putStrLn "Avaliação registrada com sucesso."

-- Função para calcular a média das avaliações
calculateAverageRating :: IO (Float, Float)
calculateAverageRating = do
    contents <- readFile "reviews.txt"
    let reviews = lines contents
        parseRating line = case words line of
            (_:ratingStr:_) -> readMaybe ratingStr :: Maybe Float
            _ -> Nothing
        infraRatings = [r | line <- reviews, "Infraestrutura:" `isInfixOf` line, Just r <- [parseRating line]]
        foodRatings = [r | line <- reviews, "Comida:" `isInfixOf` line, Just r <- [parseRating line]]
        infraAvg = if null infraRatings then 0 else sum infraRatings / fromIntegral (length infraRatings)
        foodAvg = if null foodRatings then 0 else sum foodRatings / fromIntegral (length foodRatings)
    return (infraAvg, foodAvg)

File: LOGICO(Prolog)/ReviewManagement.pl
:- use_module(library(readutil)).
:- use_module(library(files)).
:- use_module(library(lists)).

% Função para deixar uma avaliação
leave_review :-
    write("Avalie a infraestrutura da sala (nota de 0 a 10):"), nl,
    read(InfrastructureRating),
    write("Avalie a comida (nota de 0 a 10):"), nl,
    read(FoodRating),
    format(string(ReviewEntry), "Infraestrutura: ~w | Comida: ~w~n", [InfrastructureRating, FoodRating]),
    open('reviews.txt', append, Stream),
    write(Stream, ReviewEntry),
    close(Stream),
    write("Avaliação registrada com sucesso."), nl.

% Função para calcular a media das avaliações
calculate_average_rating(InfraAvg, FoodAvg) :-
    read_file_to_string('reviews.txt', Content, []),
    split_string(Content, "\n", "", Reviews),
    findall(Rating, (member(Line, Reviews), sub_string(Line, _, _, _, "Infraestrutura:"), parse_rating(Line, Rating)), InfraRatings),
    findall(Rating, (member(Line, Reviews), sub_string(Line, _, _, _, "Comida:"), parse_rating(Line, Rating)), FoodRatings),
    ( InfraRatings \= [] -> sum_list(InfraRatings, InfraSum), length(InfraRatings, InfraCount), InfraAvg is InfraSum / InfraCount ; InfraAvg is 0 ),
    ( FoodRatings \= [] -> sum_list(FoodRatings, FoodSum), length(FoodRatings, FoodCount), FoodAvg is FoodSum / FoodCount ; FoodAvg is 0 ).

% Função auxiliar para parsear a avaliação
parse_rating(Line, Rating) :-
    split_string(Line, " ", "", [_ | [RatingStr | _]]),
    atom_number(RatingStr, Rating).